cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -311,7 +312,9 @@ public void performRunLevelStop(HttpRequest request, HttpResponder responder,
                                   @PathParam("run-id") String runId,
                                   @QueryParam("graceful") String gracefulShutdownSecs) throws Exception {
     ProgramType programType = getProgramType(type);
-    ProgramId program = new ApplicationId(namespaceId, appId).program(programType, programId);
+    // Pass in the latest version to ensure the programID exists in later check
+    ProgramId program = lifecycleService.getLatestApplicationId(new ApplicationReference(namespaceId, appId))
```

Comment:
instead of fetching latest version here, we can also create a new method lifecycleService.stop(programRef, runId, gracefulShutdownSecsInt); which fetch the latest version then call lifecycleService.stop(programId,...)